teos-common/src/lightning/ser_utils.rs

```diff
@@ -0,0 +1,200 @@
+impl<'a, T: Writeable> Writeable for LightningVecWriter<'a, T> {
+    #[inline]
+    fn write<W: Writer>(&self, writer: &mut W) -> Result<(), Error> {
+        for item in self.0.iter() {
+            item.write(writer)?;
+        }
+        Ok(())
+    }
+}
+
+/// Essentially std::io::Take but a bit simpler and with a method to walk the underlying stream
+/// forward to ensure we always consume exactly the fixed length specified.
+pub(super) struct FixedLengthReader<R: Read> {
+    read: R,
+    bytes_read: u64,
+    total_bytes: u64,
+}
+
+impl<R: Read> FixedLengthReader<R> {
+    /// Returns a new FixedLengthReader.
+    pub fn new(read: R, total_bytes: u64) -> Self {
+        Self {
+            read,
+            bytes_read: 0,
+            total_bytes,
+        }
+    }
+
+    /// Returns whether there are remaining bytes or not.
+    #[inline]
+    pub fn bytes_remain(&mut self) -> bool {
+        self.bytes_read != self.total_bytes
+    }
+
+    /// Consume the remaining bytes.
+    #[inline]
+    pub fn eat_remaining(&mut self) -> Result<(), DecodeError> {
+        copy(self, &mut sink()).unwrap();
+        if self.bytes_read != self.total_bytes {
+            Err(DecodeError::ShortRead)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl<R: Read> Read for FixedLengthReader<R> {
+    #[inline]
+    fn read(&mut self, dest: &mut [u8]) -> Result<usize, Error> {
+        if self.total_bytes == self.bytes_read {
+            Ok(0)
+        } else {
+            let read_len = core::cmp::min(dest.len() as u64, self.total_bytes - self.bytes_read);
+            match self.read.read(&mut dest[0..(read_len as usize)]) {
+                Ok(v) => {
+                    self.bytes_read += v as u64;
+                    Ok(v)
+                }
+                Err(e) => Err(e),
+            }
+        }
+    }
+}
+
+/// A Read which tracks whether any bytes have been read at all. This allows us to distinguish
+/// between "EOF reached before we started" and "EOF reached mid-read".
+pub(super) struct ReadTrackingReader<R: Read> {
+    read: R,
+    /// Tells whether we have read from this reader or not yet.
+    pub have_read: bool,
+}
+
+impl<R: Read> ReadTrackingReader<R> {
+    /// Returns a new ReadTrackingReader.
+    pub fn new(read: R) -> Self {
+        Self {
+            read,
+            have_read: false,
+        }
+    }
+}
+
+impl<R: Read> Read for ReadTrackingReader<R> {
+    #[inline]
+    fn read(&mut self, dest: &mut [u8]) -> Result<usize, Error> {
+        match self.read.read(dest) {
+            Ok(0) => Ok(0),
+            Ok(len) => {
+                self.have_read = true;
+                Ok(len)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test_utils {
+    use crate::appointment::{Locator, LOCATOR_LEN};
+    use crate::cryptography::get_random_bytes;
+    use bitcoin::hashes::Hash;
+    use bitcoin::Txid;
+    pub use lightning::util::test_utils::TestVecWriter;
```

Comment:
> I'm guessing this is because you're grouping it as part of the serialization tools and then re-importing it in messages.rs and ser_macros.rs

Yup.